drivers: sensor: lis2mdl: Add single mode operation

MasoudShiroei opened the pull request:
Support single mode operation by enabling it and
making the driver to use the interrupt to findout
when the data is ready for fetch.
The sample fetch will be blocked for a specified
maximum time untill the interrupt happens.

Signed-off-by: Masoud Shiroei <masoud.shiroei@assaabloy.com>

avisconti commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -179,6 +184,18 @@ static int lis2mdl_sample_fetch_mag(const struct device *dev)
 
+#ifdef CONFIG_LIS2MDL_SINGLE_MODE
+	if (lis2mdl_operating_mode_set(lis2mdl->ctx, LIS2MDL_SINGLE_TRIGGER)) {
+		LOG_DBG("set single mode failed\n");
```
I think that here should be LOG_ERR(). 
I know that the code is full of LOG_DBG() in same type of error cases. I guess we need to fix it sooner or later.

avisconti commented on drivers/sensor/lis2mdl/Kconfig:
```diff
@@ -11,6 +11,20 @@ menuconfig LIS2MDL
+	  Specify the type of operation mode to be used by driver.
+
+config LIS2MDL_SINGLE_MODE
+	bool "Single mode"
```
Is it true that in single case mode the trigger should be put to on?
I saw that you are giving a semaphore inside the trigger.c module

MasoudShiroei replied:
If you mean lis2mdl_trigger_set() function by trigger? The answer is NO. 
But we need an interrupt handler to give the semaphore I used for blocking the fetch function. 

Therefore I used "k_sem_give(&lis2mdl->fetch_sem);" in the lis2mdl_handle_interrupt(). The lis2mdl_handle_interrupt() function will be enabled by choosing the CONFIG_LIS2MDL_TRIGGER_OWN_THREAD.

So in my implementation, the user needs to choose both the **LIS2MDL_SINGLE_MODE** and **CONFIG_LIS2MDL_TRIGGER_OWN_THREAD** to have everything working.

avisconti commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -351,12 +368,29 @@ static int lis2mdl_init(const struct device *dev)
+	if (lis2mdl_drdy_on_pin_set(lis2mdl->ctx, 1)) {
+		LOG_DBG("set drdy on pin failed!\n");
+		return -EIO;
+	}
```
The setting of drdy_on_pin is already done calling lis2mdl_trigger_set() from API.
Shouldn't we do the same even in this case?

MasoudShiroei replied:
The lis2mdl_trigger_set() will be called by the user if he/she wants to attach an **external** trigger to the interrupt. So the user is **free** to call lis2mdl_trigger_set() or **not** when it comes to "single-mode operation+interrupt". So we need to call the lis2mdl_drdy_on_pin_set() somewhere else to be sure that this is set since it is a crucial part of the "single-mode operation+interrupt".

avisconti reviewed (changes requested):
The implementation looks promising, but at a first glance I feel that there should be a better integration with trigger handling. Please let me know your feelings.

gmarull commented on drivers/sensor/lis2mdl/Kconfig:
```diff
@@ -11,6 +11,20 @@ menuconfig LIS2MDL
+config LIS2MDL_SINGLE_MODE
+	bool "Single mode"
+
+config LIS2MDL_CONTINUOUS_MODE
```
I think such setting should be made part of DeviceTree configuration. In case multiple instances are to be supported, only a single operation mode would be supported.

avisconti replied:
I felt the same yesterday. Thanks to highllight this.

gmarull commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -179,6 +184,18 @@ static int lis2mdl_sample_fetch_mag(const struct device *dev)
+#ifdef CONFIG_LIS2MDL_SINGLE_MODE
+	if (lis2mdl_operating_mode_set(lis2mdl->ctx, LIS2MDL_SINGLE_TRIGGER)) {
+		LOG_DBG("set single mode failed\n");
+		return -EIO;
```
Error code from `lis2mdl_operating_mode_set` should be propagated, even though I see current driver doesn't do that in multiple areas.

gmarull commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -351,12 +368,29 @@ static int lis2mdl_init(const struct device *dev)
+#ifdef CONFIG_LIS2MDL_SINGLE_MODE
+	/* Set device in single mode */
+	if (lis2mdl_operating_mode_set(lis2mdl->ctx, LIS2MDL_SINGLE_TRIGGER)) {
+		LOG_DBG("set single mode failed\n");
```
`LOG_ERR`

gmarull commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -179,6 +184,18 @@ static int lis2mdl_sample_fetch_mag(const struct device *dev)
+	}
+
+	if (k_sem_take(&lis2mdl->fetch_sem, K_MSEC(SAMPLE_FETCH_TIMEOUT_MS))) {
+		/* SAMPLE FETCH FAILED */
```
```suggestion

```

gmarull reviewed (commented):
Left some comments

gmarull reviewed (changes requested):
When moving parameters to DeviceTree driver should ideally be able to operate in all modes (in case multiple instances have to be supported). With these changes the driver will only support one operating mode, depending on what gets defined in `lis2mdl.h` header file.

gmarull commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -174,21 +182,79 @@ static int lis2mdl_attr_set(const struct device *dev,
-		LOG_DBG("Failed to read sample");
+	rc = lis2mdl_operating_mode_set(lis2mdl->ctx, LIS2MDL_SINGLE_TRIGGER);
+	if (rc) {
+		LOG_ERR("set single mode failed\n");
```
all `\n` should be removed from `LOG_XXX`, they're not needed

gmarull commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -174,21 +182,79 @@ static int lis2mdl_attr_set(const struct device *dev,
+	}
+
+	if (k_sem_take(&lis2mdl->fetch_sem, K_MSEC(SAMPLE_FETCH_TIMEOUT_MS))) {
+		LOG_ERR("Magnetometer data is not ready in %d ms\n",
```
```suggestion
		LOG_ERR("Magnetometer data not ready within %d ms\n",
```

gmarull commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -294,8 +360,11 @@ static int lis2mdl_init(const struct device *dev)
+	int rc = 0;
 
 	lis2mdl->dev = dev;
+	lis2mdl->single_mode = IS_ENABLED(DT_INST_PROP(0, single_mode));
```
these settings should be stored directly in driver config (`dev->config`) so that they will have the right value on each instance.

gmarull commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -373,18 +485,28 @@ static int lis2mdl_init(const struct device *dev)
 	if (new_state == DEVICE_PM_ACTIVE_STATE) {
-		status = lis2mdl_operating_mode_set(lis2mdl->ctx,
-				LIS2MDL_CONTINUOUS_MODE);
+
```
nit: extra blank line not needed?

gmarull reviewed (commented):
a few nitpicks, other than that LGTM

MasoudShiroei commented on dts/bindings/sensor/st,lis2mdl-i2c.yaml:
```diff
@@ -17,3 +17,17 @@ properties:
+
+    cancel-offset:
+      type: boolean
+      required: false
```
Yes, If I change it to true, then it is required to be set in .dts file. So it will always be enabled which is not my intention. 

If the user obviates it, the driver instance will not cancel the offset.

MasoudShiroei commented on dts/bindings/sensor/st,lis2mdl-i2c.yaml:
```diff
@@ -17,3 +17,17 @@ properties:
+
+    single-mode:
+      type: boolean
+      required: false
```
Yes, If I change it to true, then it is required to be set in .dts file. So it will always be enabled which is not my intention. 

If the user obviates it, the driver instance will work in continuous mode.

gmarull replied:
if removed from bindings, it should take `false` by default

MasoudShiroei replied:
If the user does not define it in the .dts file, it will be considered as false (0) by default. 

https://docs.zephyrproject.org/latest/guides/dts/bindings.html#bindings-file-syntax
#   - 'type: boolean' is for properties used as flags that don't take a value,
#     like
#
#         hw-flow-control;
#
#     The macro generated for the property gets set to 1 if the property exists
#     on the node, and to 0 otherwise. When combined with 'required: true',
#     this type just forces the flag to appear on the node. The output will
#     always be 1 in that case.

MasoudShiroei commented:
Please see my answers.

MaureenHelm commented:
> The implementation looks promising, but at a first glance I feel that there should be a better integration with trigger handling. Please let me know your feelings.

I agree

mbolivar-nordic commented on dts/bindings/sensor/st,lis2mdl-i2c.yaml:
```diff
@@ -17,3 +17,15 @@ properties:
+      type: boolean
+      description: |
+        Set to config the sensor in single measurement mode. Leave
+        unset to configure the sensor in continious measurement mode.
```
```suggestion
        unset to configure the sensor in continuous measurement mode.
```

MasoudShiroei commented:
> The implementation looks promising, but at a first glance I feel that there should be a better integration with trigger handling. Please let me know your feelings.

Could you please be more specific here? I do not really know what changes should I make here!

MasoudShiroei commented:
@MaureenHelm  Does any other modifications need to be included in this PR? If so, please detail them.

avisconti commented on dts/bindings/sensor/st,lis2mdl-i2c.yaml:
```diff
@@ -17,3 +17,15 @@ properties:
+      type: boolean
+      description: |
+        Set to enable the offset cancellation. Otherwise it would be
+        disabled as default.
```
Why did you change only i2c and not spi binding?

avisconti replied:
Ah! And here it would be nice to have a common st,lis2mdl-common.yaml file, to be shared between i2c and spi. You can see that there are currently already other examples.

avisconti replied:
This part seems addressed correctly now.

avisconti commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -307,13 +382,13 @@ static int lis2mdl_init(const struct device *dev)
 
 	if (wai != LIS2MDL_ID) {
-		LOG_DBG("Invalid chip ID: %02x\n", wai);
+		LOG_DBG("Invalid chip ID: %02x", wai);
```
I think all of this changes are ok, but please do it in a separate commit first.
Like "fixing all LOG messages". And on top of that commit do this one that is introducing the SINGLE mode

avisconti replied:
Well, actually are LOG_ERR(). But in any case do it in a preliminary fixing commit.

avisconti replied:
I saw that you are changing all LOG_DBG in LOG_ERR in a separate commit, as requested.
But, are you really changing ALL the LOG_DBG in the driver? They seem to me only some of them.
Am I wrong? THis is the only part missing, than for me is ok.

avisconti commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -174,21 +182,83 @@ static int lis2mdl_attr_set(const struct device *dev,
-	/* fetch raw data sample */
-	if (lis2mdl_magnetic_raw_get(lis2mdl->ctx, raw_mag) < 0) {
-		LOG_DBG("Failed to read sample");
+	rc = lis2mdl_operating_mode_set(lis2mdl->ctx, LIS2MDL_SINGLE_TRIGGER);
```
I would do that in init() function, based on single_mode is on or not. Does it make sense to you?

MasoudShiroei replied:
I tested it. It does not work in the init function. It should be called every time we want to read in the single-mode.

avisconti replied:
Ah, yes. It seems legit!

MasoudShiroei committed:
drivers: sensor: lis2mdl: Fix all log messages

Remove unnecessary "\n" at the end of logs.

Change LOG_DBG to LOG_ERR wherever is appropriate

Signed-off-by: Masoud Shiroei <masoud.shiroei@assaabloy.com>

MasoudShiroei committed:
drivers: sensor: lis2mdl: Add single mode operation

Support single mode operation by enabling it and
making the driver to use the interrupt to findout
when the data is ready for fetch. The sample fetch
will be blocked for a specified maximum time untill
the interrupt happens.

* Make operation mode configurable in DTS file
* Make offset cancellation configurable in DTS file
* Use single common .yaml file for both i2c and spi
* Store above configurations in dev->config_info

Signed-off-by: Masoud Shiroei <masoud.shiroei@assaabloy.com>

mbolivar-nordic reviewed (changes requested):
Use DT_INST_PROP instead of DT_INST_NODE_HAS_PROP for boolean properties.

mbolivar-nordic commented on drivers/sensor/lis2mdl/lis2mdl.c:
```diff
@@ -261,6 +331,9 @@ static int lis2mdl_init_interface(const struct device *dev)
 
 static const struct lis2mdl_config lis2mdl_dev_config = {
 	.master_dev_name = DT_INST_BUS_LABEL(0),
+	.single_mode = DT_INST_NODE_HAS_PROP(0, single_mode),
+	.cancel_offset = DT_INST_NODE_HAS_PROP(0, cancel_offset),
```
More details (I wanted to put the -1 in as quickly as I could to avoid this getting merged):

DT_INST_NODE_HAS_PROP should never be used for boolean properties. DT_PROP (and the DT_INST_PROP convenience helper) do the right thing with booleans:

https://docs.zephyrproject.org/latest/reference/devicetree/api.html#c.DT_PROP

```
For properties whose bindings have the following types, this macro expands to:
[...]
    boolean: 0 if the property is false, or 1 if it is true
```

mbolivar-nordic reviewed (approved):
Thanks for your patience.
